tools/depcheck/pkg/cmd/trace/trace.go

```diff
@@ -0,0 +1,428 @@
+	}
+
+	if len(o.Analyze) > 0 {
+		return o.analyzeGraph(g)
+	}
+
+	return nil
+}
+
+func (o *TraceImportsOpts) outputGraph(g graph.Directed) error {
+	if o.OutputFormat != "dot" {
+		return fmt.Errorf("invalid output format provided: %s", o.OutputFormat)
+	}
+
+	data, err := dot.Marshal(g, fmt.Sprintf("%q", strings.Join(o.Roots, ", ")), "", " ", false)
+	if err != nil {
+		return err
+	}
+
+	fmt.Fprintf(o.Out, "%v\n", string(data))
+	return nil
+}
+
+// analyzeGraph receives a MutableDirectedGraph and outputs
+// - "Yours": a list of every node in the set unique to the dependency tree of a given target node.
+// - "Mine": a list of every node in the set unique to the dependency tree of the root nodes (set non-overlapping with given target node)
+// - "Ours": a list of every node in the overlapping set between the dependency tree of the root nodes and a given target node
+func (o *TraceImportsOpts) analyzeGraph(g *depgraph.MutableDirectedGraph) error {
+	nodeToAnalyze, exists := g.NodeByName(o.Analyze)
+	if !exists {
+		return fmt.Errorf("unable to find package with import path %q", o.Analyze)
+	}
+	you, ok := nodeToAnalyze.(*depgraph.Node)
+	if !ok {
+		return fmt.Errorf("expected node to analyze to be of type *depgraph.Node. Got: %v", you)
+	}
+
+	fmt.Printf("Analyzing a total of %v packages\n", len(g.Nodes()))
+	fmt.Println()
+
+	// determine existing orphans and roots
+	keep := []graph.Node{}
+	roots := []graph.Node{}
+	for _, n := range g.Nodes() {
+		node, ok := n.(*depgraph.Node)
+		if !ok {
+			continue
+		}
+
+		to := g.To(n)
+		from := g.From(n)
+
+		// found an existing orphaned node to exclude
+		if len(to) == 0 && len(from) == 0 {
+			continue
+		}
+
+		// found a root
+		if len(from) > 0 && len(to) == 0 {
+			roots = append(roots, n)
+			continue
+		}
+
+		// skip package we are analyzing
+		if node.UniqueName == o.Analyze {
+			continue
+		}
+
+		keep = append(keep, n)
+	}
+
+	yoursGraph, err := copyGraph(g)
+	if err != nil {
+		return err
+	}
+	yours, err := printYours(yoursGraph, roots, keep, you)
+	if err != nil {
+		return err
+	}
+	fmt.Println()
+
+	// no need to copy the graph a second time - we can mutate it
+	// since it won't be used anymore
+	mine, err := printMine(g, roots, keep, you)
+	if err != nil {
+		return err
+	}
+
+	ours := []graph.Node{}
+	yoursOrMine := append(yours, mine...)
+	for _, n := range keep {
+		isYoursOrMine := false
+		for _, yom := range yoursOrMine {
+			if n.ID() == yom.ID() {
+				isYoursOrMine = true
+				break
+			}
+		}
+		if !isYoursOrMine {
+			ours = append(ours, n)
+		}
+	}
+	fmt.Println()
+
+	fmt.Printf("\"Ours\": %v packages shared between the main repo and %q\n", len(ours), you.UniqueName)
+	for _, n := range ours {
+		pkg, ok := n.(*depgraph.Node)
+		if !ok {
+			continue
+		}
+
+		fmt.Printf("    - %s\n", pkg.UniqueName)
+	}
+
+	return nil
+}
+
+func printYours(g *depgraph.MutableDirectedGraph, roots, nodes []graph.Node, you *depgraph.Node) ([]graph.Node, error) {
```

Comment:
The logic in here looks wrong.  I expected you to produce a new graph by filtering out a selected set of nodes.  Then after you've built your new graph, you simply scan all nodes to see if they don't have any inbound links. You track those, then built a new graph using the nodes without any inbound links as your input.  Repeat until you have no changes.  Then union all the nodes you removed.

Reply:
Notice that at no point do we discuss "roots".  I'm still trying to figure out why that concept is needed.